ops/model.py

```diff
@@ -797,6 +1001,314 @@ def __init__(self, name: str, address_info: '_AddressDict'):
+                raise RuntimeError(
+                    "{} has no revision, but is owned by this unit.".format(self)
+                )  # should not happen in production.
+            else:
+                return None
+        return self._revision
+
+    def __str__(self):
+        label = " <?>"
+        if self.label:
+            label = " label={!r}".format(self.label)
+        revision = "<?>"
+        if self.revision:
+            revision = " revision={!r}".format(self.revision)
+        return '<{}{}{} owned-by-this-unit={}>'.format(self.id, revision, label,
+                                                       self._is_owned_by_this_unit)
+
+    @classmethod
+    def _ensure_id(cls, id: Optional[str], operation: str = "do this on") -> str:
+        if not id:  # empty string will fail the regex match afterwards anyway
+            raise ValueError(
+                'Cannot {} a secret with ID={!r}. If you created this secret via '
+                '`self.model.get_secret(label=X)`, you should pass `id` instead.'
+                ''.format(id, operation))
+        if not isinstance(id, str):
+            raise TypeError(id)
+
+        id = id.strip()  # trim any whitespace
+
+        # add the prefix if not there already
+        if not id.startswith(cls._secret_id_prefix):
+            id = cls._secret_id_prefix + id
+
+        if not cls._secret_id_re.match(id):
+            raise ValueError("{!r} is not a valid secret id.".format(id))
+
+        return id
+
+    @classmethod
+    def _validate_content(cls, content: Optional[Dict[str, str]]):
+        if not content:
+            # fixme: ATM you can't secret-set without content (e.g. to only update the label),
+            #  but that is a bug in juju.
+            #  when that is fixed, `secret-set [id] --label X` will be valid.
+            raise ModelError('missing secret value or filename')
+
+        invalid_secret_keys = [key for key in content if not cls._secret_key_re.match(key)]
+        invalid_secret_values = [val for val in content.values() if not isinstance(val, str)]  # type: List[Any]  # noqa
+        if invalid_secret_keys or invalid_secret_values:
+            msg = "Secret content invalid."
+            if invalid_secret_keys:
+                msg += '\nInvalid keys: {}. Secret content keys should be ' \
+                       'at least 3 characters long, ' \
+                       'no number to start, no dash to end.'.format(invalid_secret_keys)
+            if invalid_secret_values:
+                bad_types = [type(obj).__name__ for obj in invalid_secret_values]
+                msg += '\nInvalid values: {}. Secret content values should be of ' \
+                       'type str, not {}.'.format(invalid_secret_values, bad_types)
+            raise ValueError(msg)
+
+    def _validate_ownership(self, operation_name: str,
+                            should_own: bool = True):
+        owns = self._is_owned_by_this_unit
```

Comment:
So checking - this "is owned" value is set if the current unit is the leader (for app owned secrets)?